src/server/snapshot/snapshotter.ts

```diff
@@ -86,33 +91,46 @@ export class Snapshotter {
       }
       this._delegate.onFrameSnapshot(snapshot);
     });
-    const initScript = '(' + frameSnapshotStreamer.toString() + ')()';
-    await this._context._doAddInitScript(initScript);
-    for (const page of this._context.pages())
-      await page.mainFrame()._evaluateExpression(initScript, false, undefined, 'main');
   }
 
   dispose() {
     helper.removeEventListeners(this._eventListeners);
   }
 
-  async forceSnapshot(page: Page, snapshotId: string) {
-    await Promise.all([
-      page.frames().forEach(async frame => {
-        try {
-          const context = await frame._mainContext();
-          await context.evaluateInternal(({ kSnapshotStreamer, snapshotId }) => {
-            // Do not block action execution on the actual snapshot.
-            Promise.resolve().then(() => (window as any)[kSnapshotStreamer].forceSnapshot(snapshotId));
-            return undefined;
-          }, { kSnapshotStreamer, snapshotId });
-        } catch (e) {
-        }
-      })
-    ]);
+  async captureSnapshot(page: Page, snapshotId: string) {
+    const snapshotFrame = async (frame: Frame) => {
+      const context = frame._existingMainContext();
+      await context?.evaluateInternal(({ kSnapshotStreamer, snapshotId }) => {
+        (window as any)[kSnapshotStreamer].captureSnapshot(snapshotId);
+        return undefined;
```

Comment:
nit: do we need the `return undefined` statements? (a few in the PR).

Reply:
Not in this case, but typically we make sure that if we don't need the return value, return undefined prevents it from being serialized.